### ISSUE TITLE
Design Proposal: Benchmark Configs on Openjdk-tests Framework

### ISSUE THREAD

**Issue by piyush286** [AdoptOpenJDK/openjdk-tests#853]
### Background of the Current Design for Running Benchmarks through Openjdk-tests Framework

In order to integrate all performance benchmarks into openjdk-tests repo, we need to add benchmark configurations for several benchmarks and variants that are able to run on all supported HW platforms by OpenJDK. As part of this initiative, we tested the water by deploying ODM benchmark on Openjdk-tests Framework since that benchmark is simple to configure and run, though it needs a license. 

Initially, we designed the Openjdk-tests Framework to use a bash script (i.e. `benchmark.sh` as shown below) that would keep all the bash commands, which are required to set environment variables and run the benchmark. 

https://github.com/AdoptOpenJDK/openjdk-tests/blob/0cff000dabf262f48ad0c20f45f36892f0220fcb/performance/odm/scripts/benchmark.sh#L1-L55

### Limitations of the Current Design

Some issues with keeping the benchmark configs as export commands inside a static bash script:

1. This design is not very flexible or maintainable if we want to run benchmarks with multiple variations. 

2. Environment Variables are hard-coded for just one machine right now. So, you would need to manually change environment variables such as `CLIENT`, ‘DB_MACHINE`, `LIBERTY_HOST` and `AFFINITY` while running some benchmarks, especially the ones that require more than one machine, such as Liberty DayTrader. You would also need to change the command for pinning threads such as numactl.

3. Redundancy! Currently, all performance tests can be run from the PerfNext launcher (https://github.com/AdoptOpenJDK/openjdk-test-tools/tree/master/PerfNext/), which already has all the benchmark configurations in that repo. 

If you see the ODM configs file from PerfNext below, you'll notice that those configs have been duplicated in the openjdk-tests repo for now. The only major difference is the format of configs. On openjdk-tests, we use a bash script that has all the configs in export commands, while on PerfNext, we keep the configs in a XML file.

https://github.com/AdoptOpenJDK/openjdk-test-tools/blob/b528a63ea146080f4bb1430740a734c2af87df7e/PerfNext/config/benchmarks/data_simple/ODM.xml#L1-L55

### Possible Solutions

In order to avoid having the same benchmark configurations in multiple locations, we should use the configs from the PerfNext repo. Currently, `playlist.xml` (shown below) uses the static bash script (i.e. `benchmark.sh`). `playlist.xml` should call some tool/script to dynamically generate that bash script using the benchmark configs from PerfNext and then run that generated script.

https://github.com/AdoptOpenJDK/openjdk-tests/blob/0cff000dabf262f48ad0c20f45f36892f0220fcb/performance/odm/playlist.xml#L18-L24

### Option 1:
PerfNext also has an API (snippet below) that can generate the bash script for running a benchmark with a specific name and variant. It can also generate the specific HW environment variables such as CPU affinity commands and client machine for a selected machine. So we can possibly use that PerfNext's feature (`/api/benchengine/generatescript`) to generate the bash script for running benchmarks through Openjdk-tests Framework.

Code snippet from https://github.com/AdoptOpenJDK/openjdk-test-tools/blob/b528a63ea146080f4bb1430740a734c2af87df7e/PerfNext/app/apis/BenchEngine/parser.js#L20-L34

```
	/*
    * API to generate test script from XML Benchmark definition.     
    * 
    * Params: XML definition
    */
	app.post('/api/benchengine/generatescript', function(req, res) {
		var raw_definition = req.rawBody; //Raw XML to be parsed

		parseDefinition(raw_definition, function(generatedOutput){
			parseSetupScript(raw_definition, function(setupScript){
				var mergedScripts = setupScript + generatedOutput;
				res.send(mergedScripts)
			});
		});
	});
```

### Option 2:
1. Create an parameter for Openjdk-tests Framework to specify the location of the benchmark configs XML file (i.e. https://github.com/AdoptOpenJDK/openjdk-test-tools/tree/master/PerfNext/config) in case we need to get these files from some other location (i.e. internal) in future.
2. Openjdk-tests Framework should download the XML files from the specified location 
3. Write a simple script that parses the benchmark XML files for general benchmark configs and machine database XML file for machine specific configs for some specific benchmark variant, and generates a bash script that can set all the parameters and run the benchmark.

### Option 3:
Combined design of option 1 & option 2.

### Assigned Contributors
Awsaf (@pinicman) from my team would work on adding this functionality.

**Comment by piyush286** [comment]
@llxia @ShelleyLambert We are thinking of what might be the best option out of those mentioned above for using benchmark configs on Openjdk-tests Framework. Would you guys have some suggestions?

**Comment by smlambert** [comment]
This design will take time to flesh out.  As we have already discussed, we have identified that the config files in PerfNext could potentially be collapsed into the same format (with some additions) to the format as the playlist files.

I think it is important to distinguish between the 2 main use cases:

1. the nightly runs, which do not need to be dynamically generated, they are config files with default settings in them

1. the runs that a person using the PerfNext GUI to generate a config file, so overwriting default settings

Option 4 (essentially its your Option 2, but moving the script generation down to lower layer, as you no longer require it in Perfnext if you are calling tkg):

For 1) playlist file contains the test targets for the nightly build, that links to a default config file that is checked into the performance/subdir folder for a particular benchmark  (in this way the nightlies do not depend at all on Perfnext instance to run, it is NOT acceptable to depend on Perfnext to be running, in order for tkg to run)

For 2), playlist file contains a test target called custom_benchmarkName, your are using Perfnext GUI, which generates a custom config that calls custom_benchmarkName target with a path or link to the location of the custom config file

Do not keep the logic for generating benchmark scripts (what Salman named BenchEngine) in Perfnext.  It can be decoupled, moved to the lower layer.  You avoid duplication, and simplify Perfnext, which essentially should be as dumb a GUI as possible.  (as in, most if not all functionality should be possible without the GUI layer present).

We can discuss in a web conference soon, if this wasn't clear.

**Comment by piyush286** [comment]
Some of the work in this issue might be related to [Enable PerfNext To Use Openjdk-tests Framework](https://github.com/AdoptOpenJDK/openjdk-test-tools/issues/31).

**Comment by piyush286** [comment]
Thanks Shelley for the suggestions.

Jotting down some more thoughts that came up so that I don't forget. We can discuss these thoughts as well in our meeting.

> 1. the nightly runs, which do not need to be dynamically generated, they are config files with default settings in them

Totally agree! 

> Do not keep the logic for generating benchmark scripts (what Salman named BenchEngine) in Perfnext. It can be decoupled, moved to the lower layer. You avoid duplication, and simplify Perfnext, which essentially should be as dumb a GUI as possible. (as in, most if not all functionality should be possible without the GUI layer present).

`/api/benchengine/submit` API currently sends 2 scripts to Jenkins Pipeline:
setupScript: It does all the setup tasks such as downloading benchmark and SDK packages.
benchmarkScript: Exports all the necessary environment variables and runs the main benchmark script.

We could get rid of the `setupScript` since Openjdk-tests Framework can already do that. We would still need the ability to generate `benchmarkScript` so that developers are able to change default configs and PerfNext can pass the custom benchmark script to Openjdk-tests Framework to run.

**Comment by smlambert** [comment]
Yes, let's keep listing the requirements, and not get bogged down in how to accomplish it yet.  It is important to clearly know "what needs to be accomplished" before getting distracted by "how it will be implemented".
